pkg/controllers/subnet/subnet_controller.go

```diff
@@ -50,102 +48,144 @@ type SubnetReconciler struct {
 }
 
 func (r *SubnetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	obj := &v1alpha1.Subnet{}
-	log.Info("reconciling subnet CR", "subnet", req.NamespacedName)
+	startTime := time.Now()
+	defer func() {
+		log.Info("Finished reconciling Subnet", "Subnet", req.NamespacedName, "duration", time.Since(startTime))
+	}()
 	metrics.CounterInc(r.SubnetService.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnet)
 
-	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
-		log.Error(err, "unable to fetch Subnet CR", "req", req.NamespacedName)
-		return ResultNormal, client.IgnoreNotFound(err)
-	}
-
-	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
-		metrics.CounterInc(r.SubnetService.NSXConfig, metrics.ControllerUpdateTotal, MetricResTypeSubnet)
-		if !controllerutil.ContainsFinalizer(obj, servicecommon.SubnetFinalizerName) {
-			controllerutil.AddFinalizer(obj, servicecommon.SubnetFinalizerName)
-
-			if obj.Spec.AccessMode == "" {
-				log.Info("obj.Spec.AccessMode set", "subnet", req.NamespacedName)
-				obj.Spec.AccessMode = v1alpha1.AccessMode(v1alpha1.AccessModePrivate)
-			}
-
-			if obj.Spec.IPv4SubnetSize == 0 {
-				vpcNetworkConfig := r.VPCService.GetVPCNetworkConfigByNamespace(obj.Namespace)
-				if vpcNetworkConfig == nil {
-					err := fmt.Errorf("operate failed: cannot get configuration for Subnet CR")
-					log.Error(nil, "failed to find VPCNetworkConfig for Subnet CR", "subnet", req.NamespacedName, "namespace %s", obj.Namespace)
-					updateFail(r, ctx, obj, err.Error())
-					return ResultRequeue, err
-				}
-				obj.Spec.IPv4SubnetSize = vpcNetworkConfig.DefaultSubnetSize
-			}
-			if err := r.Client.Update(ctx, obj); err != nil {
-				log.Error(err, "add finalizer", "subnet", req.NamespacedName)
-				updateFail(r, ctx, obj, err.Error())
+	subnetCR := &v1alpha1.Subnet{}
+	if err := r.Client.Get(ctx, req.NamespacedName, subnetCR); err != nil {
+		// IgnoreNotFound returns nil on NotFound errors.
+		if client.IgnoreNotFound(err) == nil {
+			if err := r.deleteSubnetByName(req.Name, req.Namespace); err != nil {
+				log.Error(err, "failed to delete NSX Subnet", "Subnet", req.NamespacedName)
 				return ResultRequeue, err
 			}
-			log.V(1).Info("added finalizer on subnet CR", "subnet", req.NamespacedName)
+			return ResultNormal, nil
 		}
-		tags := r.SubnetService.GenerateSubnetNSTags(obj, obj.Namespace)
-		if tags == nil {
-			return ResultRequeue, errors.New("failed to generate subnet tags")
+		log.Error(err, "unable to fetch Subnet CR", "req", req.NamespacedName)
+		return ResultRequeue, err
```

Comment:
I've discussed with wenqi in this case we should requeue it since  we will get client.IgnoreNotFound(err) == nil
`if client.IgnoreNotFound(err) == nil`